src/Compilers/Core/Portable/MetadataReader/PEModule.cs

```diff
@@ -1623,26 +1621,109 @@ private bool HasStringAndIntValuedAttribute(EntityHandle token, AttributeDescrip
+        {
+            CrackObsoleteProperties(out var diagnosticId, out var urlFormat, ref sig);
+            value = new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message: null, isError: false, diagnosticId: diagnosticId, urlFormat: urlFormat);
+            return true;
+        }
+
+        /// <summary>
+        /// Cracks the data on an Obsolete attribute using the Obsolete(string) constructor.
+        /// </summary>
+        private static bool CrackObsoleteAttributeDataString(out ObsoleteAttributeData? value, ref BlobReader sig)
+        {
+            string message;
+            if (CrackStringInAttributeValue(out message, ref sig))
+            {
+                CrackObsoleteProperties(out var diagnosticId, out var urlFormat, ref sig);
+                value = new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError: false, diagnosticId: diagnosticId, urlFormat: urlFormat);
+                return true;
+            }
+
+            value = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Cracks the data on an Obsolete attribute using the Obsolete(string, bool) constructor.
+        /// </summary>
+        private static bool CrackObsoleteAttributeDataStringBool(out ObsoleteAttributeData? value, ref BlobReader sig)
         {
             string message;
             if (CrackStringInAttributeValue(out message, ref sig) && sig.RemainingBytes >= 1)
             {
                 bool isError = sig.ReadBoolean();
-                value = new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError);
+
+                CrackObsoleteProperties(out var diagnosticId, out var urlFormat, ref sig);
+                value = new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError, diagnosticId: diagnosticId, urlFormat: urlFormat);
                 return true;
             }
 
             value = null;
             return false;
         }
 
+        private static void CrackObsoleteProperties(out string? diagnosticId, out string? urlFormat, ref BlobReader sig)
+        {
+            diagnosticId = null;
+            urlFormat = null;
+
+            // See CIL spec section II.23.3 Custom attributes
+
+            // Next is a description of the optional “named” fields and properties.
+            // This starts with NumNamed– an unsigned int16 giving the number of “named” properties or fields that follow.
+            var numNamed = sig.ReadUInt16();
+            for (int i = 0; i < numNamed; i++)
+            {
+                // A NamedArg is simply a FixedArg preceded by information to identify which field or property it represents.
+                // PROPERTY is the single byte 0x54.
+                const int PROPERTY = 0x54;
+                var fieldOrProperty = sig.ReadByte();
+                if (fieldOrProperty != PROPERTY)
+                {
+                    return;
```

Comment:
>return; [](start = 20, length = 7)

This is not correct. We shouldn't bail out on things we do not expect. Instead we should pay attention to things we are looking for and skip anything else. Please add tests with fields encoded before the properties.